### DISCUSSION TITLE
Using LTSP as a base to netboot and operate servers

### DISCUSSION THREAD

**Discussion by stumbaumr** [ltsp/ltsp#377] [Q&A] [answered]
Hi,

we use Debian 10 LTSP5 as our base to operate all our servers in more than one location. We are used to PXE booting from a NetApp clustered NFS storage and are currently transitioning to a Proxmox Ceph based setup.
We will have two VMs within Proxmox running a keepalived Ganesha NFS active-passive setup handing out CephFS shares.

We are used to PXE boot from readonly NetApp .snapshot directories. So we would install/upgrade software into our sole image, take a new snapshot, adjust the DHCP entry with the new version and reboot the VMs/Servers. In case of problems we would just revert the DHCP entry to a previously working snapshot.

This should still be possible with CephFS: we will mkdir .snap/versionid on CephFS and boot from that NFS exported .snap directory.

Data and Configuration used to live on extra NFS exports and would be mounted by a bespoke script, the configuration used to contain the information which systemd services to enable & start or disable. If necessary we could also mount bind specific configuration files.
Data will live now on a virtio Disk on the specific VM/Server using ZFS, so we can mount the required dataset at the specific mountpoint and have regular snapshot backups within the VM/Server
Configuration and shared directories will still live on NFSv4 exports.

Now we have the following issue: The NFS ganesha supports NFSv4 and I would like to use that. It works fine within Debian 10 but not within the initramfs. I played around with NFSv4 previously but to no avail (see https://bugs.launchpad.net/ltsp/+bug/1510155 ).

So how should I go from here?

1. Develop my own Diskless server netboot setup?
2. Try configure Ganesha to operate NFSv3 and NFSv4 properly and leave the VM setup as is?
3. Upgrade Debian 10 with LTSP19 - would our setup even be possible?
4. Any other ideas?

Happy new year and stay healthy!
Rainer

**Comment by alkisg** [comment, marked as answer]
Hi Rainer,

you're using many subsystems that I have no experience with, so I'm not sure I can give proper advice.
Let's start a chat and others can chime in, e.g. there's a pull request for NFS ganesha in https://github.com/ltsp/ltsp/pull/88 and some pull requests for CephFS and RBD as well, e.g. https://github.com/ltsp/ltsp/issues/37.

My first question is... I didn't understand why you need to boot using NFSv4. AFAIK all NFS servers can support both v3 and v4 at the same time, so you could boot the rootfs with v3, while mounting /home and /shared/dirs over NFSv4.

Your server/snapshot handling sounds a bit confusing to me; for example, I maintain 30 different images using VirtualBox, which of course supports VM snapshots and rollback, and I can netboot any of them using the stock iPXE menu (and DEFAULT_IMAGE in ltsp.conf). LTSP even keeps the last squashfs version around, so if an update breaks things, it takes one second to run `ltsp image --revert buster-mate`, I don't even need to restore the last good snapshot and re-publish the image.  And I can even netboot the VMs without running `ltsp image vm` at all, if I'm using flat (raw) VM files. But it sounded like that part works fine for you, so no need to change your workflow there.

ZFS is great, I've used it for /home and minutely/hourly/daily/weekly snapshots without issues.

In launchpad you asked about dracut; if Debian can be properly netbooted using dracut, using a single directory for a single client (plain NFS4 netbooting), then sure we could add LTSP support for that, even without testing with Fedora etc yet. But can it? The last time I checked, dracut had multiple issues there, that would need to be resolved outside of LTSP...

Some other options could be to use CephFS+RBD for rootfs, which would also do fault tolerance, while NFS4 doesn't have that, or to use NFS3 + wireshark for super-safety and super-compatibility.

These were my initial thoughts, when I understand the problem better I'll be able to propose more stuff.

**Reply by stumbaumr** [reply]
Hi Alkis,

thanks for your quick reponse!

I started off with a default nfs-ganesha file based setup and completey lost the idea to boot from NFSv3 and mount after the initramfs using NFSv4. This is the approach I am now taking and I already have the first system booting.

We still have Debian 8 systems booting from these snapshots in various versions but most is Debian 10 by now - currently three different release versions of our image. But also three VMs being still on Debian 7 :-( . All within that single NFS export:

```
root@controlnode01.dc1:~# ls -al /export/diskless/
total 68
drwxr-xr-x 12 root root  4096 Feb 14  2020 .
drwx------ 17 root root  4096 Mar  4  2020 ..
drwxr-xr-x 17 root root  4096 Jul 24 14:15 debian10-amd64-clnt
drwxr-xr-x 28 root root  4096 Sep  7  2017 debian10-amd64-srv
drwxr-xr-x 29 root root  4096 Oct  8  2019 debian7-amd64-srv
drwxr-xr-x 23 root root  4096 Oct 10  2017 debian7-i386-clnt
drwxr-xr-x 29 root root  4096 Jan 16  2019 debian8-amd64-srv
drwxr-xr-x 22 root root  4096 Oct  8  2019 debian8-i386-clnt
drwxr-xr-x 22 root root  4096 Feb 18  2020 debian9-amd64-srv
drwxr-xr-x 27 root root  4096 Dec 14  2017 debiantst-amd64-srv
drwxr-xr-x 20 root root  4096 Aug 17  2019 gentoo-amd64-srv
drwxrwxrwx 53 root root 20480 Jan 15 00:00 .snapshot
drwxr-xr-x 16 root root  4096 Jan 14 10:04 tftproot
root@controlnode01.dc1:~# ls -al /export/diskless/debian10-amd64-srv/.snapshot/
total 204
drwxrwxrwx 53 root root 20480 Jan 15 00:00 .
drwxr-xr-x 28 root root  4096 Sep  7  2017 ..
drwxr-xr-x 28 root root  4096 Sep  7  2017 00206
drwxr-xr-x 28 root root  4096 Sep  7  2017 00207
drwxr-xr-x 28 root root  4096 Sep  7  2017 00208
drwxr-xr-x 28 root root  4096 Sep  7  2017 00209
drwxr-xr-x 28 root root  4096 Sep  7  2017 00210
drwxr-xr-x 28 root root  4096 Sep  7  2017 00211
drwxr-xr-x 28 root root  4096 Sep  7  2017 00212
drwxr-xr-x 28 root root  4096 Sep  7  2017 00213
drwxr-xr-x 28 root root  4096 Sep  7  2017 00214
drwxr-xr-x 28 root root  4096 Sep  7  2017 00215
drwxr-xr-x 28 root root  4096 Sep  7  2017 00216
drwxr-xr-x 28 root root  4096 Sep  7  2017 00217
drwxr-xr-x 28 root root  4096 Sep  7  2017 00218
drwxr-xr-x 28 root root  4096 Sep  7  2017 00219
drwxr-xr-x 28 root root  4096 Sep  7  2017 00220
drwxr-xr-x 28 root root  4096 Sep  7  2017 00221
drwxr-xr-x 28 root root  4096 Sep  7  2017 00222
root@controlnode01.dc1:~#
```
The date on the snapshot list above is not correct - snapshot  00206 was taken in 2019
```
root@controlnode01.dc1:~# ls -al /export/diskless/debian10-amd64-srv/.snapshot/00206/var/log/apt/history.log
-rw-r--r-- 1 root root 1533536 Jul 26  2019 /export/diskless/debian10-amd64-srv/.snapshot/00206/var/log/apt/history.log
root@controlnode01.dc1:~#
```

In the DHCP config we have groups per Debian Version and Network. For an upgrade from Debian 9 to Debian 10 we just move the DHCP entry of the storage network interface to the Debian 10 group and reboot.

For testing around we just boot directly from the writable mountpoint
```
root@controlnode01.dc1:~# cat /etc/dhcp/dhcpd.conf
...
                option root-path "10.30.0.3:/vol/diskless/debian10-amd64-srv/.snapshot/00222";
                filename ".snapshot/00222/debian10-amd64-srv/pxelinux.0";
                host disklesstest {
                        hardware ethernet 00:50:56:8e:01:93;
                        fixed-address 10.30.1.20;
                        #option root-path "10.30.0.2:/vol/diskless/debian10-amd64-srv";
                        #filename "debian10-amd64-srv/pxelinux.0";
                        option root-path "10.30.0.2:/vol/diskless/.snap/00222/debian10-amd64-srv";
                        filename ".snap/00222/debian10-amd64-srv/pxelinux.0";
                        }
...
```
That currently active config for disklesstest is booting from a CephFS snapshot handed out via NFS Ganesha :-)

So I think I went a bit into panic mode yesterday when I was not able to boot as NFSv3 was not active on that Ganesha NFS service...

Best regards and thanks for pushing in me in the correct direction!
Rainer
